ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -24,35 +27,149 @@
+ * +---------+----------------------+-----------+-------------+-------------+----------+----------+
+ * </pre>
  */
 public class BlobId extends StoreKey {
-  private Short version = 1;
-  private static short Version_Size_In_Bytes = 2;
-  private PartitionId partitionId;
-  private String uuid;
-  private static int UUID_Size_In_Bytes = 4;
+  public static final byte DEFAULT_FLAG = 0;
+  // version 1 of the serialized format
+  static final short BLOB_ID_V1 = 1;
+  // version 2 of the serialized format
+  static final short BLOB_ID_V2 = 2;
+  private static final short CURRENT_VERSION = BLOB_ID_V1;
+  private static final short VERSION_FIELD_LENGTH_IN_BYTES = Short.BYTES;
+  private static final short UUID_SIZE_FIELD_LENGTH_IN_BYTES = Integer.BYTES;
+  private static final short FLAG_FIELD_LENGTH_IN_BYTES = Byte.BYTES;
+  private static final short DATACENTER_ID_FIELD_LENGTH_IN_BYTES = Byte.BYTES;
+  private static final short ACCOUNT_ID_FIELD_LENGTH_IN_BYTES = Short.BYTES;
+  private static final short CONTAINER_ID_FIELD_LENGTH_IN_BYTES = Short.BYTES;
+  // the version to indicate the serialized format.
+  private final Short version;
+  private final Byte flag;
+  private final Byte datacenterId;
+  private final Short accountId;
+  private final Short containerId;
+  private final PartitionId partitionId;
+  protected final String uuid;
+
+  /**
+   * Constructs a new BlobId by taking arguments for the required fields. The constructed BlobId will be serialized
+   * into {@link #CURRENT_VERSION}. If {@code CURRENT_VERSION == BLOB_ID_V1}, it will serialize itself into
+   * {@code BLOB_ID_V1}, ignoring {@code flag}, {@code datacenterId}, {@code containerId}, and {@code containerId}
+   * regardless they are set or not.
+   * @param flag A byte to embed additional information of this blobId. Will be reset to {@link #DEFAULT_FLAG} if
+   *             {@link #CURRENT_VERSION} is {@link #BLOB_ID_V1}.
+   * @param datacenterId The id of the datacenter to be embedded into the blob. Will be reset to
+   *             {@link ClusterMapUtils#UNKNOWN_DATACENTER_ID} if {@link #CURRENT_VERSION} is {@link #BLOB_ID_V1}.
+   * @param accountId The id of the {@link Account} to be embedded into the blob. Will be reset to
+   *             {@link Account#UNKNOWN_ACCOUNT_ID} if {@link #CURRENT_VERSION} is {@link #BLOB_ID_V1}.
+   * @param containerId The id of the {@link Container} to be embedded into the blob. Will be reset to
+   *             {@link Container#UNKNOWN_CONTAINER_ID} if {@link #CURRENT_VERSION} is {@link #BLOB_ID_V1}.
+   * @param partitionId The partition where this blob is to be stored. Cannot be {@code null}.
+   */
+  public BlobId(byte flag, byte datacenterId, short accountId, short containerId, PartitionId partitionId) {
+    if (partitionId == null) {
+      throw new IllegalArgumentException("partitionId cannot be null");
+    }
+    version = getCurrentVersion();
```

Comment:
I know I had suggested overriding `getCurrentVersion()` to test V2, but I did not mean for this to be called from the constructor. Constructors invoking overridable methods is generally considered bad form. `BlobIdV2` in the tests that extends `BlobId` is not completely constructed when the superclass constructor here calls into `BlobIdV2::getCurrentVersion()`. If this method was accessing object state then it would lead to undefined behavior (was okay here because it just returns a `static final` variable)

This can be avoided by doing this check within the other methods (that is instead of checking `if (version == x)`, those other methods would do `if (getCurrentVersion() == x)`.